src/libusb.c

```diff
@@ -0,0 +1,153 @@
+#define BENDPOINTADDRESS_IN             0x80
+#define TYPICAL_NIC_PACKET_SIZE         0x0200
+#define MAX_ENDPOINTS                   1000
+
+static bool
+libusb_is_ethernet_interface(struct usb_interface_descriptor *interface)
+{
+  return (interface->bInterfaceClass == COMMUNICATIONS_CLASS &&
+    interface->bInterfaceSubClass == ETHERNET_NETWORKING_SUBCLASS);
+}
+
+static bool
+libusb_is_wireless_interface(struct usb_interface_descriptor *interface)
+{
+  return (interface->bInterfaceClass == WIRELESS_CLASS &&
+    (interface->bInterfaceSubClass != RADIO_FREQUENCY_SUBCLASS ||
+    interface->bInterfaceProtocol != BLUETOOTH_PROTOCOL));
+}
+
+static bool
+libusb_is_bluetooth_interface(struct usb_interface_descriptor *interface)
+{
+  return (interface->bInterfaceClass == WIRELESS_CLASS &&
+    interface->bInterfaceSubClass == RADIO_FREQUENCY_SUBCLASS &&
+    interface->bInterfaceProtocol == BLUETOOTH_PROTOCOL);
+}
+
+/**
+ * Identify vendor-specific interfaces with endpoints that support
+ * bulk data transfers both in and out with a packet size of at least
+ * TYPICAL_NIC_PACKET_SIZE (512). Though all the USB NICs that were
+ * tested had such an endpoint and none of the other USB devices did,
+ * It is possible that checking for NICs in this manner will result
+ * in false positives or negatives.
+ */
+static bool
+libusb_is_unknown_bulk_transfer_interface(struct usb_interface_descriptor *interface)
+{
+  int i;
+  bool has_bulk_in = false;
+  bool has_bulk_out = false;
+
+  if (interface->bInterfaceClass == VENDOR_SPECIFIC_CLASS)
+    for (i = 0; i < interface->bNumEndpoints; i++)
+    {
+      struct usb_endpoint_descriptor *endpoint = &interface->endpoint[i];
+
+      if (endpoint->bmAttributes == BMATTRIBUTES_BULK && endpoint->wMaxPacketSize >= TYPICAL_NIC_PACKET_SIZE)
+      {
+        if (endpoint->bEndpointAddress & BENDPOINTADDRESS_IN)
+          has_bulk_in = true;
+        else
+          has_bulk_out = true;
+        if (has_bulk_in && has_bulk_out)
+          return true;
+      }
+    }
+  return false;
+}
+
+static struct usb_device *
+libusb_findDevice(int vendorid, int productid)
+{
+  struct usb_bus *bus;
+  struct usb_device *dev;
+  struct usb_bus *busses;
+
+  usb_find_busses();
+  usb_find_devices();
+  busses = usb_get_busses();
+
+  for (bus = busses; bus; bus = bus->next)
+    for (dev = bus->devices; dev; dev = dev->next)
+      if ((dev->descriptor.idVendor == vendorid) && (dev->descriptor.idProduct == productid))
+        return dev;
+
+  return NULL;
+}
+
+/**
+ * Determine if the device is a possible NIC or Bluetooth device.
+ * It parses the USB descriptors, which can have multiple
+ * configurations, which can have multiple interfaces, which can have
+ * multiple "altSettings," which can have multiple endpoints.
+ */
+void
+libusb_find_more_about_nic(device_t *device)
+{
+  int i, j, k;
+  int totalEndpoints = 0;
+  struct usb_device *libusb_device = libusb_findDevice(device->vendorid, device->deviceid);	
+
```

Comment:
You should probably check that you didn't get NULL back from libusb_findDevice before use.